File: internal/gamekeys/gamekeys.go
package gamekeys

import (
	"context"
	"df-hud/internal/config"
	"df-hud/internal/desktop"
	"df-hud/internal/model"
	"log"
	"sync"
	"sync/atomic"
	"time"
)

// Pressing a key in the game's own window, once per launch.
//
// There is exactly one thing this is for: the client's FPS readout starts OFF
// every single launch. It is not a setting that got lost - the game keeps its
// settings in the Wine registry under Software\Creaky Corpse\Dead Frontier, and
// that key has entries for the FPS CAP, vsync, quality, brightness and the HUD
// outlines, but none for whether the readout is showing. There is nothing to
// preset, so the only way to have it on is to press the key.
//
// SilverOverlays does the same thing, and has to focus the game's window first
// because it types with pynput, which goes wherever focus is. Addressing the
// window instead means nothing is focused and nothing is stolen - see
// hyprClient.SendKey.
//
// Deliberately narrow. This is not a macro facility: it sends ONE configured key
// ONCE per game process, it never repeats, and it does nothing at all while the
// launcher is the only window open.

// Sender is the compositor, as an interface so the driver can be tested
// without Hyprland.
type Sender interface {
	SendKey(ctx context.Context, key, address string) error
}

// Keys fires the configured key when the game's own window has been up long
// enough.
type Keys struct {
	cfg   func() *config.Config
	state func() model.GameState
	place func() desktop.Placement
	send  Sender

	// active is the focused window's address. A key sent to an unfocused window
	// is the likeliest reason one silently does nothing - SilverOverlays calls
	// activate() first for the same reason. This waits for focus rather than
	// stealing it. Nil means "assume focused".
	active func(context.Context) (string, bool)

	// ready reports that the client is in the world rather than on a loading
	// screen. Nil, or true, means send anyway.
	ready func() bool

	// enabled and dismiss are the live copies behind the tray checkboxes. The app
	// persists tray changes to config.toml and ApplyConfig updates them on reload.
	enabled atomic.Bool
	dismiss atomic.Bool

	mu sync.Mutex
	// pid is the game process the fields below describe. Everything resets when
	// it changes, which is what makes these once per LAUNCH rather than once per
	// df-hud lifetime.
	pid    int
	seenAt time.Time
	sent   bool
	// The launcher dialog is a separate one-shot with its own clock: it happens
	// before the game window exists, so it cannot share the other one's.
	launcherSeenAt time.Time
	launcherSent   bool
}

func New(cfg func() *config.Config, state func() model.GameState, place func() desktop.Placement, send Sender) *Keys {
	k := &Keys{cfg: cfg, state: state, place: place, send: send}
	k.enabled.Store(cfg().GameKeys.FPSDisplay)
	k.dismiss.Store(cfg().GameKeys.DismissLauncher)
	return k
}

// SetFPSDisplay is the tray menu's switch.
//
// Turning it on mid-session sends the key straight away, because that is what
// ticking a box called "FPS display" is asking for. Turning it off sends nothing
// - the game's own key is a toggle and df-hud does not track its state, so
// "off" can only mean "do not do this at the next launch". Ticking it twice in
// one session does not send twice, which is what stops that difference from
// turning into a toggle that fights the player.
func (k *Keys) SetFPSDisplay(on bool) { k.enabled.Store(on) }

func (k *Keys) FPSDisplay() bool { return k.enabled.Load() }

// ApplyConfig makes live config reloads authoritative over the two tray-backed
// switches. Tray clicks persist these fields before changing the runtime state.
func (k *Keys) ApplyConfig(cfg config.GameKeysConfig) {
	k.enabled.Store(cfg.FPSDisplay)
	k.dismiss.Store(cfg.DismissLauncher)
}

// KeysTick is the floor on how long the launcher dialog stays on screen: it
// is noticed on one tick and dismissed on a later one. Reads cached state only;
// the compositor is asked just before a key is sent. SilverOverlays uses 100ms.
const KeysTick = 200 * time.Millisecond

// Run drives the check until ctx ends. A ticker rather than an event
// subscription: the trigger is "the window has existed for N, and is focused,
// and the client is in the world", which is a clock question with two gates.
func (k *Keys) Run(ctx context.Context) {
	ticker := time.NewTicker(KeysTick)
	defer ticker.Stop()
	for {
		select {
		case <-ctx.Done():
			return
		case now := <-ticker.C:
			k.Tick(ctx, now)
		}
	}
}

// tick is the whole decision, taking the clock so it can be tested without
// waiting for one.
func (k *Keys) Tick(ctx context.Context, now time.Time) {
	game := model.GameState{}
	if k.state != nil {
		game = k.state()
	}
	if !game.Running {
		k.forget()
		return
	}
	cfg := k.cfg()
	place := k.place()

	// The launcher dialog, while it is the only thing open. Pressing its default
	// button is the whole feature; see GameKeysConfig.DismissLauncher for why it
	// is dismissed rather than stopped from appearing.
	if place.LauncherOnly && place.LauncherAddress != "" {
		k.dismissLauncher(ctx, cfg, game.PID, place.LauncherAddress, now)
		return
	}

	// The launcher is not the game. place.Known is false both while only the
	// configuration dialogs are open and while the real window is not mapped yet,
	// and either way there is nothing to send a key to - so this doubles as the
	// "wait until the client is actually up" gate.
	if !place.Known || place.Address == "" {
		return
	}

	k.mu.Lock()
	if k.pid != game.PID {
		k.pid, k.seenAt, k.sent = game.PID, now, false
		k.launcherSeenAt, k.launcherSent = time.Time{}, false
	}
	if k.seenAt.IsZero() {
		// First sight of this launch's WINDOW. Separate from the pid check above
		// because dismissing the launcher claims the pid while the game window
		// does not exist yet - and without this, that would leave the delay
		// measured from the zero time, so the key would fire on the first tick
		// after the window appeared instead of fps_delay later.
		//
		// The delay has to run from here rather than from the process starting:
		// the process starts when the LAUNCHER opens, which can sit on screen
		// indefinitely.
		k.seenAt = now
	}
	due := !k.sent && k.enabled.Load() && !now.Before(k.seenAt.Add(cfg.GameKeys.FPSDelay.Duration))
	k.mu.Unlock()
	if !due {
		return
	}
	// The two gates below are WAITS, not failures: nothing is claimed, so the
	// next tick tries again. A key pressed at the loading screen or into another
	// window is simply lost, and losing it silently is what made this
	// intermittent.
	if !k.canSend(ctx, place.Address) {
		return
	}

	k.mu.Lock()
	// Claimed before the send, not after: a send that fails has still had its one
	// attempt, and retrying against a refusing compositor would be a keypress
	// storm aimed at the game.
	k.sent = true
	k.mu.Unlock()

	key := cfg.GameKeys.FPSKey
	if err := k.send.SendKey(ctx, key, place.Address); err != nil {
		log.Printf("game keys: could not send %q to the game window: %v", key, err)
		return
	}
	log.Printf("game keys: sent %q to turn the game's FPS display on", key)
}

// SetDismissLauncher is the tray menu's switch, and the reason it is on the menu
// at all: the dialog is where input bindings are changed, so "let it stay up this
// time" has to be reachable without editing a file. Unticking it before a launch
// is the whole workflow.
func (k *Keys) SetDismissLauncher(on bool) { k.dismiss.Store(on) }

func (k *Keys) DismissLauncher() bool { return k.dismiss.Load() }

// dismissLauncher presses the dialog's default button, once per launch.
func (k *Keys) dismissLauncher(ctx context.Context, cfg *config.Config, pid int, address string, now time.Time) {
	if !k.dismiss.Load() {
		return
	}
	k.mu.Lock()
	if k.pid != pid {
		k.pid, k.seenAt, k.sent = pid, time.Time{}, false
		k.launcherSeenAt, k.launcherSent = now, false
	}
	if k.launcherSeenAt.IsZero() {
		k.launcherSeenAt = now
	}
	due := !k.launcherSent && !now.Before(k.launcherSeenAt.Add(cfg.GameKeys.LauncherDelay.Duration))
	k.mu.Unlock()
	if !due {
		return
	}
	// Focus only. The client-in-world gate does not apply: the launcher IS what
	// stands between here and the world, so waiting for the world would deadlock.
	if k.active != nil {
		if addr, ok := k.active(ctx); !ok || addr != address {
			return
		}
	}

	k.mu.Lock()
	k.launcherSent = true
	k.mu.Unlock()

	key := cfg.GameKeys.LauncherKey
	if err := k.send.SendKey(ctx, key, address); err != nil {
		log.Printf("game keys: could not dismiss the launcher: %v", err)
		return
	}
	log.Printf("game keys: pressed %q on the launcher dialog", key)
}

// canSend is the two conditions that decide whether a key would actually land:
// the client is past its loading screen, and the window it is aimed at is the
// focused one.
func (k *Keys) canSend(ctx context.Context, address string) bool {
	if k.ready != nil && !k.ready() {
		return false
	}
	if k.active == nil {
		return true
	}
	addr, ok := k.active(ctx)
	return ok && addr == address
}

// forget re-arms for the next launch.
func (k *Keys) forget() {
	k.mu.Lock()
	k.pid, k.seenAt, k.sent = 0, time.Time{}, false
	k.launcherSeenAt, k.launcherSent = time.Time{}, false
	k.mu.Unlock()
}

func (k *Keys) SetActive(fn func(context.Context) (string, bool)) { k.active = fn }
func (k *Keys) SetReady(fn func() bool)                           { k.ready = fn }

File: src/gamekeys.rs
//! Press one configured key in the game window, once per launch.
//!
//! The FPS readout starts off every launch and is not a saved setting, so the
//! only way to have it on is to press the game's own key. The launcher dialog
//! is a separate one-shot so the Input tab stays reachable by unticking the
//! tray box.

use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{Arc, Mutex};
use std::time::{Duration, SystemTime};

use crate::config::GameKeys as GameKeysCfg;
use crate::desktop::{Client, Placement};
use crate::model::GameState;

pub struct Keys {
    fps: AtomicBool,
    dismiss: AtomicBool,
    inner: Mutex<Inner>,
}

#[derive(Default)]
struct Inner {
    pid: i32,
    seen_at: Option<SystemTime>,
    sent: bool,
    launcher_seen_at: Option<SystemTime>,
    launcher_sent: bool,
}

impl Keys {
    pub fn new(cfg: &GameKeysCfg) -> Arc<Self> {
        let k = Self {
            fps: AtomicBool::new(cfg.fps_display),
            dismiss: AtomicBool::new(cfg.dismiss_launcher),
            inner: Mutex::new(Inner::default()),
        };
        Arc::new(k)
    }

    pub fn set_fps_display(&self, on: bool) {
        self.fps.store(on, Ordering::SeqCst);
    }

    pub fn fps_display(&self) -> bool {
        self.fps.load(Ordering::SeqCst)
    }

    pub fn set_dismiss_launcher(&self, on: bool) {
        self.dismiss.store(on, Ordering::SeqCst);
    }

    pub fn dismiss_launcher(&self) -> bool {
        self.dismiss.load(Ordering::SeqCst)
    }

    pub fn apply_config(&self, cfg: &GameKeysCfg) {
        self.fps.store(cfg.fps_display, Ordering::SeqCst);
        self.dismiss.store(cfg.dismiss_launcher, Ordering::SeqCst);
    }

    pub fn tick(
        &self,
        now: SystemTime,
        cfg: &GameKeysCfg,
        game: GameState,
        place: &Placement,
        send: &dyn Sender,
        active: Option<&str>,
        ready: bool,
    ) {
        if !game.running {
            *self.inner.lock().unwrap() = Inner::default();
            return;
        }

        if place.launcher_only && !place.launcher_address.is_empty() {
            self.dismiss_launcher_tick(now, cfg, game.pid, &place.launcher_address, send, active);
            return;
        }
        if !place.known || place.address.is_empty() {
            return;
        }

        let due = {
            let mut g = self.inner.lock().unwrap();
            if g.pid != game.pid {
                g.pid = game.pid;
                g.seen_at = Some(now);
                g.sent = false;
                g.launcher_seen_at = None;
                g.launcher_sent = false;
            }
            if g.seen_at.is_none() {
                g.seen_at = Some(now);
            }
            !g.sent
                && self.fps.load(Ordering::SeqCst)
                && now
                    .duration_since(g.seen_at.unwrap())
                    .unwrap_or(Duration::ZERO)
                    >= cfg.fps_delay.0
        };
        if !due {
            return;
        }
        if !can_send(&place.address, active, ready) {
            return;
        }

        self.inner.lock().unwrap().sent = true;
        if let Err(err) = send.send_key(&cfg.fps_key, &place.address) {
            eprintln!(
                "game keys: could not send {:?} to the game window: {err}",
                cfg.fps_key
            );
            return;
        }
        eprintln!(
            "game keys: sent {:?} to turn the game's FPS display on",
            cfg.fps_key
        );
    }

    fn dismiss_launcher_tick(
        &self,
        now: SystemTime,
        cfg: &GameKeysCfg,
        pid: i32,
        address: &str,
        send: &dyn Sender,
        active: Option<&str>,
    ) {
        if !self.dismiss.load(Ordering::SeqCst) {
            return;
        }
        let due = {
            let mut g = self.inner.lock().unwrap();
            if g.pid != pid {
                g.pid = pid;
                g.seen_at = None;
                g.sent = false;
                g.launcher_seen_at = Some(now);
                g.launcher_sent = false;
            }
            if g.launcher_seen_at.is_none() {
                g.launcher_seen_at = Some(now);
            }
            !g.launcher_sent
                && now
                    .duration_since(g.launcher_seen_at.unwrap())
                    .unwrap_or(Duration::ZERO)
                    >= cfg.launcher_delay.0
        };
        if !due {
            return;
        }
        if !can_send(address, active, true) {
            return;
        }

        self.inner.lock().unwrap().launcher_sent = true;
        if let Err(err) = send.send_key(&cfg.launcher_key, address) {
            eprintln!("game keys: could not dismiss the launcher: {err}");
            return;
        }
        eprintln!(
            "game keys: pressed {:?} on the launcher dialog",
            cfg.launcher_key
        );
    }
}

fn can_send(address: &str, active: Option<&str>, ready: bool) -> bool {
    ready && active == Some(address)
}

pub trait Sender {
    fn send_key(&self, key: &str, address: &str) -> Result<(), String>;
}

impl Sender for dyn Client {
    fn send_key(&self, key: &str, address: &str) -> Result<(), String> {
        Client::send_key(self, key, address)
    }
}

impl Sender for Box<dyn Client> {
    fn send_key(&self, key: &str, address: &str) -> Result<(), String> {
        Client::send_key(self.as_ref(), key, address)
    }
}

pub fn spawn(handle: Arc<crate::app::Handle>, stop: Arc<std::sync::atomic::AtomicBool>) {
    std::thread::Builder::new()
        .name("df-hud-gamekeys".into())
        .spawn(move || {
            let send = crate::desktop::new_client();
            while !stop.load(Ordering::SeqCst) && !handle.stopped() {
                let cfg = handle.cfg.lock().unwrap().game_keys.clone();
                let game = handle.game.state();
                let place = handle.vis.placement();
                let ready = handle.store.client_in_world(chrono::Utc::now());
                let active = send.active_address();
                handle.gamekeys.tick(
                    SystemTime::now(),
                    &cfg,
                    game,
                    &place,
                    &send,
                    active.as_deref(),
                    ready,
                );
                std::thread::sleep(Duration::from_millis(200));
            }
        })
        .ok();
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::Duration as CfgDuration;
    use crate::config::GameKeys as GameKeysCfg;
    use std::sync::Mutex;
    use std::time::Duration;

    struct Fake {
        sent: Mutex<Vec<(String, String)>>,
        err: Mutex<Option<String>>,
    }

    impl Fake {
        fn new() -> Self {
            Self {
                sent: Mutex::new(Vec::new()),
                err: Mutex::new(None),
            }
        }
        fn count(&self) -> usize {
            self.sent.lock().unwrap().len()
        }
        fn last(&self) -> (String, String) {
            self.sent.lock().unwrap().last().cloned().unwrap()
        }
    }

    impl Sender for Fake {
        fn send_key(&self, key: &str, address: &str) -> Result<(), String> {
            self.sent
                .lock()
                .unwrap()
                .push((key.to_string(), address.to_string()));
            match self.err.lock().unwrap().as_ref() {
                Some(e) => Err(e.clone()),
                None => Ok(()),
            }
        }
    }

    fn cfg() -> GameKeysCfg {
        GameKeysCfg {
            fps_display: true,
            fps_key: "y".into(),
            fps_delay: CfgDuration(Duration::from_secs(5)),
            dismiss_launcher: false,
            launcher_key: "Return".into(),
            launcher_title: "Dead Frontier Configuration".into(),
            launcher_delay: CfgDuration(Duration::from_millis(500)),
        }
    }

    fn running(pid: i32, address: &str) -> (GameState, Placement) {
        (
            GameState {
                running: true,
                pid,
                ..GameState::default()
            },
            Placement {
                known: true,
                address: address.into(),
                ..Placement::default()
            },
        )
    }

    fn launcher(pid: i32, address: &str) -> (GameState, Placement) {
        (
            GameState {
                running: true,
                pid,
                ..GameState::default()
            },
            Placement {
                launcher_only: true,
                launcher_address: address.into(),
                ..Placement::default()
            },
        )
    }

    fn tick(
        k: &Keys,
        now: SystemTime,
        cfg: &GameKeysCfg,
        game: GameState,
        place: &Placement,
        send: &Fake,
    ) {
        let dest = if !place.address.is_empty() {
            Some(place.address.as_str())
        } else if !place.launcher_address.is_empty() {
            Some(place.launcher_address.as_str())
        } else {
            None
        };
        k.tick(now, cfg, game, place, send, dest, true);
    }

    #[test]
    fn fps_key_is_sent_once_the_window_has_been_up_for_the_delay() {
        let k = Keys::new(&cfg());
        let send = Fake::new();
        let start = SystemTime::UNIX_EPOCH + Duration::from_secs(1000);
        let (game, place) = running(42, "0xabc");
        tick(&k, start, &cfg(), game, &place, &send);
        assert_eq!(send.count(), 0);
        tick(
            &k,
            start + Duration::from_secs(4),
            &cfg(),
            game,
            &place,
            &send,
        );
        assert_eq!(send.count(), 0);
        tick(
            &k,
            start + Duration::from_secs(5),
            &cfg(),
            game,
            &place,
            &send,
        );
        assert_eq!(send.count(), 1);
        assert_eq!(send.last(), ("y".into(), "0xabc".into()));
    }

    #[test]
    fn fps_key_is_sent_only_once_per_launch() {
        let k = Keys::new(&cfg());
        let send = Fake::new();
        let start = SystemTime::UNIX_EPOCH + Duration::from_secs(1000);
        let (game, place) = running(42, "0xabc");
        tick(&k, start, &cfg(), game, &place, &send);
        for i in 0..30 {
            tick(
                &k,
                start + Duration::from_secs(5 + i),
                &cfg(),
                game,
                &place,
                &send,
            );
        }
        assert_eq!(send.count(), 1);
    }

    #[test]
    fn fps_key_waits_out_the_launcher() {
        let k = Keys::new(&cfg());
        let send = Fake::new();
        let start = SystemTime::UNIX_EPOCH + Duration::from_secs(1000);
        let (game, place) = launcher(42, "");
        for i in 0..60 {
            tick(
                &k,
                start + Duration::from_secs(i),
                &cfg(),
                game,
                &place,
                &send,
            );
        }
        assert_eq!(send.count(), 0);
        let window = start + Duration::from_secs(60);
        let (game, place) = running(42, "0xabc");
        tick(&k, window, &cfg(), game, &place, &send);
        assert_eq!(send.count(), 0);
        tick(
            &k,
            window + Duration::from_secs(5),
            &cfg(),
            game,
            &place,
            &send,
        );
        assert_eq!(send.count(), 1);
    }

    #[test]
    fn fps_key_rearms_when_the_pid_changes() {
        let k = Keys::new(&cfg());
        let send = Fake::new();
        let start = SystemTime::UNIX_EPOCH + Duration::from_secs(1000);
        let (game, place) = running(42, "0xabc");
        tick(&k, start, &cfg(), game, &place, &send);
        tick(
            &k,
            start + Duration::from_secs(5),
            &cfg(),
            game,
            &place,
            &send,
        );
        let second = start + Duration::from_secs(6);
        let (game, place) = running(99, "0xdef");
        tick(&k, second, &cfg(), game, &place, &send);
        tick(
            &k,
            second + Duration::from_secs(5),
            &cfg(),
            game,
            &place,
            &send,
        );
        assert_eq!(send.count(), 2);
        assert_eq!(send.last().1, "0xdef");
    }

    #[test]
    fn fps_key_sends_nothing_when_off() {
        let k = Keys::new(&cfg());
        k.set_fps_display(false);
        let send = Fake::new();
        let start = SystemTime::UNIX_EPOCH + Duration::from_secs(1000);
        let (game, place) = running(42, "0xabc");
        for i in 0..20 {
            tick(
                &k,
                start + Duration::from_secs(i),
                &cfg(),
                game,
                &place,
                &send,
            );
        }
        assert_eq!(send.count(), 0);
    }

    #[test]
    fn fps_key_fires_when_turned_on_mid_session_but_not_twice() {
        let k = Keys::new(&cfg());
        k.set_fps_display(false);
        let send = Fake::new();
        let start = SystemTime::UNIX_EPOCH + Duration::from_secs(1000);
        let (game, place) = running(42, "0xabc");
        tick(&k, start, &cfg(), game, &place, &send);
        tick(
            &k,
            start + Duration::from_secs(30),
            &cfg(),
            game,
            &place,
            &send,
        );
        assert_eq!(send.count(), 0);
        k.set_fps_display(true);
        tick(
            &k,
            start + Duration::from_secs(31),
            &cfg(),
            game,
            &place,
            &send,
        );
        assert_eq!(send.count(), 1);
        k.set_fps_display(false);
        tick(
            &k,
            start + Duration::from_secs(32),
            &cfg(),
            game,
            &place,
            &send,
        );
        k.set_fps_display(true);
        for i in 0..10 {
            tick(
                &k,
                start + Duration::from_secs(33 + i),
                &cfg(),
                game,
                &place,
                &send,
            );
        }
        assert_eq!(send.count(), 1);
    }

    #[test]
    fn fps_key_does_not_retry_a_failed_send() {
        let k = Keys::new(&cfg());
        let send = Fake::new();
        *send.err.lock().unwrap() = Some("window not found".into());
        let start = SystemTime::UNIX_EPOCH + Duration::from_secs(1000);
        let (game, place) = running(42, "0xabc");
        tick(&k, start, &cfg(), game, &place, &send);
        for i in 0..20 {
            tick(
                &k,
                start + Duration::from_secs(5 + i),
                &cfg(),
                game,
                &place,
                &send,
            );
        }
        assert_eq!(send.count(), 1);
    }

    #[test]
    fn launcher_is_dismissed_once_after_the_delay() {
        let mut cfg = cfg();
        cfg.dismiss_launcher = true;
        let k = Keys::new(&cfg);
        k.set_dismiss_launcher(true);
        let send = Fake::new();
        let start = SystemTime::UNIX_EPOCH + Duration::from_secs(1000);
        let (game, place) = launcher(42, "0x111");
        tick(&k, start, &cfg, game, &place, &send);
        assert_eq!(send.count(), 0);
        tick(
            &k,
            start + Duration::from_secs(1),
            &cfg,
            game,
            &place,
            &send,
        );
        assert_eq!(send.count(), 1);
        assert_eq!(send.last(), ("Return".into(), "0x111".into()));
        for i in 2..20 {
            tick(
                &k,
                start + Duration::from_secs(i),
                &cfg,
                game,
                &place,
                &send,
            );
        }
        assert_eq!(send.count(), 1);
    }

    #[test]
    fn launcher_is_left_alone_when_off() {
        let k = Keys::new(&cfg());
        let send = Fake::new();
        let start = SystemTime::UNIX_EPOCH + Duration::from_secs(1000);
        let (game, place) = launcher(42, "0x111");
        for i in 0..30 {
            tick(
                &k,
                start + Duration::from_secs(i),
                &cfg(),
                game,
                &place,
                &send,
            );
        }
        assert_eq!(send.count(), 0);
    }

    #[test]
    fn fps_delay_runs_from_the_window_after_dismissing_the_launcher() {
        let mut cfg = cfg();
        cfg.dismiss_launcher = true;
        let k = Keys::new(&cfg);
        k.set_dismiss_launcher(true);
        let send = Fake::new();
        let start = SystemTime::UNIX_EPOCH + Duration::from_secs(1000);
        let (game, place) = launcher(42, "0x111");
        tick(&k, start, &cfg, game, &place, &send);
        tick(
            &k,
            start + Duration::from_secs(1),
            &cfg,
            game,
            &place,
            &send,
        );
        assert_eq!(send.count(), 1);
        let window = start + Duration::from_secs(3);
        let (game, place) = running(42, "0xabc");
        tick(&k, window, &cfg, game, &place, &send);
        tick(
            &k,
            window + Duration::from_secs(4),
            &cfg,
            game,
            &place,
            &send,
        );
        assert_eq!(send.count(), 1);
        tick(
            &k,
            window + Duration::from_secs(5),
            &cfg,
            game,
            &place,
            &send,
        );
        assert_eq!(send.count(), 2);
        assert_eq!(send.last(), ("y".into(), "0xabc".into()));
    }

    #[test]
    fn fps_display_is_off_by_default() {
        let cfg = crate::config::Config::default();
        assert!(!cfg.game_keys.fps_display);
        assert_eq!(cfg.game_keys.fps_key, "y");
        assert!(!cfg.game_keys.dismiss_launcher);
    }
}
